app/services/catalog/create_approval_request.rb

```diff
@@ -0,0 +1,62 @@
+module Catalog
+  class CreateApprovalRequest
+    attr_reader :order
+
+    def initialize(id)
+      @order = Order.find_by!(:id => id)
+    end
+
+    def process
+      @order.order_items.each do |order_item|
+        approvals = send_approval_requests(order_item)
+        approvals.compact.each do |req|
+          order_item.approval_requests << create_approval_request(req)
+        end
+      end
+
+      @order.update(:state => "Ordered", :ordered_at => Time.now.utc)
+      self
+    rescue Catalog::ApprovalError => e
+      Rails.logger.error("Error putting in approval Request for #{order.id}: #{e.message}")
+      raise
+    end
+
+    private
+
+    def send_approval_requests(order_item)
+      workflows(order_item.portfolio_item).each_with_object([]) do |workflow, approvals|
+        Approval.call do |api_instance|
+          approvals << api_instance.create_request(workflow, request_body_from(order_item))
+        end
+        order_item.update_message("info", "Approval Request Submitted for workflow #{workflow}")
+      end
+    end
+
+    def request_body_from(order_item)
+      svc_params_sanitized = Catalog::OrderItemSanitizedParameters.new(:order_item_id => order_item.id).process
+
+      ApprovalApiClient::RequestIn.new.tap do |request|
+        request.name      = order_item.portfolio_item.name
+        request.content   = {
+          :product   => order_item.portfolio_item.name,
+          :portfolio => order_item.portfolio_item.portfolio.name,
+          :order_id  => order_item.order_id,
+          :params    => svc_params_sanitized.to_json
+        }
+      end
+    end
+
+    def workflows(portfolio_item)
+      portfolio_item.resolved_workflow_refs
+    end
+
+    def create_approval_request(req)
+      ApprovalRequest.new.tap do |approval|
+        approval.workflow_ref         = req.workflow_id
+        approval.approval_request_ref = req.id
+        approval.state                = req.decision.to_sym
+        approval.save
```

Comment:
Do we want this to be `save!`? The listener would log this case if the `save` here failed so I'm not actually sure. But it feels like if we can't save this, then the process should fail.

Reply:
good catch, otherwise it would have juts blown right through and not warned.